// zenrows/dist/index.js
// src/index.ts
import fastq from "fastq";
import fetchRetry from "fetch-retry";

// package.json
var package_default = {
  name: "zenrows",
  version: "1.3.2",
  description: "ZenRows Node SDK",
  main: "./dist/index.cjs",
  module: "./dist/index.js",
  types: "./dist/index.d.ts",
  engines: {
    node: ">=20"
  },
  packageManager: "pnpm@9.4.0",
  exports: {
    ".": {
      require: "./dist/index.cjs",
      import: "./dist/index.js",
      types: "./dist/index.d.ts"
    }
  },
  scripts: {
    build: "tsup src/index.ts --dts --format cjs,esm --clean",
    test: "vitest"
  },
  keywords: ["sdk", "zenrows", "scraping"],
  author: "ZenRows",
  repository: {
    type: "git",
    url: "https://github.com/ZenRows/zenrows-node-sdk.git"
  },
  homepage: "https://github.com/ZenRows/zenrows-node-sdk#readme",
  url: "https://github.com/ZenRows/zenrows-node-sdk/issues",
  type: "module",
  license: "MIT",
  devDependencies: {
    "@biomejs/biome": "1.8.1",
    "@types/async-retry": "^1.4.8",
    "@types/node": "^20.14.2",
    msw: "^2.3.1",
    tsup: "^8.1.0",
    typescript: "^5.4.5",
    vitest: "^1.6.0"
  },
  dependencies: {
    fastq: "^1.17.1",
    "fetch-retry": "^6.0.0"
  }
};

// src/index.ts
var API_URL = "https://api.zenrows.com/v1/";
var ZenRows = class {
  apiKey;
  clientConfig;
  queue;
  fetchWithRetry;
  constructor(apiKey, clientConfig = {}) {
    this.apiKey = apiKey;
    this.clientConfig = clientConfig;
    const retries = this.clientConfig.retries ?? 0;
    this.queue = fastq.promise(
      this,
      this.worker,
      this.clientConfig.concurrency ?? 5
    );
    this.fetchWithRetry = fetchRetry(fetch, {
      retryDelay: (attempt) => 2 ** attempt * 1e3,
      // retryOn: [422, 503, 504],
      retryOn: (attempt, error, response) => {
        if (attempt > retries) {
          return false;
        }
        if (error !== null || response?.status === 422 || response?.status === 503 || response?.status === 504) {
          return true;
        }
        return false;
      }
    });
  }
  get(url, config, { headers = {} } = {}) {
    return this.queue.push({ url, config, headers });
  }
  post(url, config, { headers = {}, data = {} } = {
    headers: { "Content-Type": "application/x-www-form-urlencoded" }
  }) {
    return this.queue.push({
      url,
      method: "POST",
      config,
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        ...headers
      },
      data
    });
  }
  async worker({
    url,
    method = "GET",
    config,
    headers,
    data
  }) {
    const params = new URLSearchParams({
      url,
      apikey: this.apiKey
    });
    if (config) {
      for (const [key, value] of Object.entries(config)) {
        if (value !== void 0) {
          params.append(key, String(value));
        }
      }
    }
    if (headers && Object.keys(headers).length) {
      params.append("custom_headers", "true");
    }
    const finalHeaders = {
      "User-Agent": `zenrows/${package_default.version} node`,
      ...headers
    };
    const fetchOptions = {
      method,
      headers: finalHeaders
    };
    if (method === "POST" && data) {
      if (typeof data === "object") {
        fetchOptions.body = JSON.stringify(data);
      } else {
        fetchOptions.body = String(data);
      }
    }
    const response = await this.fetchWithRetry(
      `${API_URL}?${params.toString()}`,
      fetchOptions
    );
    return response;
  }
};
export {
  ZenRows
};
